self-healing/transform-ldes-data-to-entities.ts

```diff
@@ -1,10 +1,94 @@
-import { querySudo } from "@lblod/mu-auth-sudo";
+import { querySudo, updateSudo } from "@lblod/mu-auth-sudo";
 import { sparqlEscapeUri } from "mu";
 
 import { HEALING_DUMP_GRAPH, HEALING_TRANSFORMED_GRAPH } from "../config";
 import { DIRECT_DB_ENDPOINT } from "../config";
 
 export async function transformLdesDataToEntities() {
+  while (await hasMoreDuplicates()) {
+    await deleteBatch();
+  }
+  await transformRemainingEntities();
+}
+
+async function hasMoreDuplicates() {
+  const result = await querySudo(`
+  PREFIX tree: <https://w3id.org/tree#>
+  PREFIX dct: <http://purl.org/dc/terms/>
+  PREFIX prov: <http://www.w3.org/ns/prov#>
+
+  SELECT * WHERE {
+    ?ldesStream tree:member ?toDelete.
+    ?toDelete dct:isVersionOf ?entity.
+    ?toDelete prov:generatedAtTime ?time.
+
+    ?ldesStream tree:member ?ldesEntityTwo.
+    ?ldesEntityTwo dct:isVersionOf ?entity.
+    ?ldesEntityTwo prov:generatedAtTime ?otherTime.
+    FILTER(?ldesEntityTwo != ?toDelete)
+
+    FILTER( ?otherTime > ?time )
+  } LIMIT 1`);
+  return result.results.bindings.length > 0;
+}
+
+async function deleteBatch() {
```

Comment:
Splitting the assign to be deleted and the delete itself is also better for virtuoso than doing it in one query? Like you've done yesterday?